service/history/workflow/mutable_state_impl.go

```diff
@@ -4378,6 +4378,16 @@ func (ms *MutableStateImpl) UpdateWorkflowStateStatus(
 func (ms *MutableStateImpl) StartTransaction(
 	namespaceEntry *namespace.Namespace,
 ) (bool, error) {
+	if ms.hBuilder.IsDirty() || len(ms.InsertTasks) > 0 {
+		ms.logger.Error("MutableState encountered dirty transaction",
+			tag.WorkflowNamespaceID(ms.executionInfo.NamespaceId),
+			tag.WorkflowID(ms.executionInfo.WorkflowId),
+			tag.WorkflowRunID(ms.executionState.RunId),
+			tag.Value(ms.hBuilder),
+		)
+		return false, serviceerror.NewInternal("MutableState encountered dirty transaction")
```

Comment:
should we return an retryable error?

Reply:
this logic should return internal since the sanity check failed. however, if the internal error will not be retried then the error type should be changed, what do you suggest?